unified flow interfaces

jkinable opened the pull request:
Added a `FlowAlgorithm` interface to ensure that flow algorithms (min cost flow/max flow) have a unified interface. The purpose of this interface is not to allow the user to swap min cost flow and max flow algorithms (this doesn't seem to be very useful). Instead, the interface ensures that all flow algorithms have similar methods and names.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/FlowAlgorithm.java:
```diff
@@ -0,0 +1,104 @@
+import java.util.*;
+
+/**
+ * Iterface for flow algorithms
```
Typo: Iterface -> Interface

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/FlowAlgorithm.java:
```diff
@@ -0,0 +1,104 @@
+    Map<E, Double> getFlowMap();
+
+    /**
+     * For the specified {@code edge} $(u, v)$ return vertex $v$ if the flow goes from $u$ to $v$, or returns
```
returns vertex $v$

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/FlowAlgorithm.java:
```diff
@@ -0,0 +1,104 @@
+     * Represents a minimum cost flow.
+     *
+     * @param <E> graph edge type
+     * @since July 2018
```
Do we still add `@since` tags to the classes?

jkinable replied:
Nah we don't, they got removed since most of the classes didn't have them.

jsichi commented on jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/FlowAlgorithm.java:
```diff
@@ -0,0 +1,111 @@
+     * For the specified {@code edge} $(u, v)$ returns vertex $v$ if the flow goes from $u$ to $v$, or returns
+     * vertex $u$ otherwise. For directed flow networks the result is always the head of the specified arc.
+     * <p>
+     * <em>Note:</em> not all minimum cost flow algorithms may support undirected graphs.
```
This should be just "not all flow algorithms"

d-michail commented:
There is one issue that needs some discussion, which is somehow related to this one.

So far our interfaces force the "output types" of the algorithms, but have no particular requirement for the input types. We tend to provide input parameters from the constructor of each class and then call the interface methods to get the result back in a well defined format. 

As long as the input is straightforward, like in shortest paths or minimum spanning trees, this is ok. What happens however when the input becomes more complicated like in the minimum cost flow problem. Do we want to also have the input parameters in the interface?  Which of the below two versions we prefer?

 - `MinimumCostFlow<E> getMinimumCostFlow()`
 - `MinimumCostFlow<E> getMinimumCostFlow(MinimumCostFlowProblem input)`

I actually like very much that Timofey has bundled all problem parameters in a well define object.

jkinable commented:
@d-michail good point, @jsichi brought this up last weekend as well. 
I would personally prefer `MinimumCostFlow<E> getMinimumCostFlow(MinimumCostFlowProblem input)` because this increases the consistency with the other classes. For instance, we have `MaximumFlow<E> getMaximumFlow(V source, V sink);` which also takes the problem params as input.

Toptachamann commented:
I agree with @jkinable: there is no need to create separate classes for ShortestPathProblem or MaxFlowProblem.  

The problem with this approach:
 * `MinimumCostFlow<E> getMinimumCostFlow(MinimumCostFlowProblem input)`
is that the class should become stateless, i.e. we no longer create an instance of the algorithm for a particular problem and then run it. In the maximum flow algorithms, we can specify the graph at construction time and then compute max. flow between different pairs of vertices. Since in many algorithms the input graph is specified at construction time, it'd be consistent to do the same thing with min. cost flow algorithm.

jkinable commented:
Just to come to a consensus:
We could have a  `MinimumCostFlowProblem` interface, similar to the current `MinimumCostFlowProblem` class, including the `getGraph` method, with a  `MinimumCostFlowProblemImpl` implementation class.
In addition we would have the interface method `getMinimumCostFlow(MinimumCostFlowProblem input)`. The graph can still be provided at construction time. Whenever the graph provided at construction time is different from the graph provided in the `MinimumCostFlowProblem input` when ` getMinimumCostFlow` is called, an IllegalArgumentException is thrown. That way, the implementation doesn't have to be stateless.

Toptachamann commented:
In general it is good to have a unifying interface method for getting the result of the algorithm. In this case, we have to add a check to the methods `getDualSolution()` and `testOptimaltiy` whether the problem has been specified already.

jkinable commented:
I've added a `MinimumCostFlowProblem` interface, and adjusted `MinimumCostFlowAlgorithm` accordingly. The `MinimumCostFlow` is now semi-stateless: each time a `getMinimumCostFlow(MinimumCostFlowProblem problem)` is invoked, the  algorithm recomputes the flow. Methods like `getDual` and `getCost` only work when `getMinimumCostFlow(MinimumCostFlowProblem problem)` has been invoked before.

@d-michail @Toptachamann have a look and let me know what you think.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/mincost/CapacityScalingMinimumCostFlow.java:
```diff
@@ -136,10 +136,6 @@
      * Computed minimum cost flow
      */
     private MinimumCostFlow<E> minimumCostFLow;
```
typo, there is a capital L in `FLow`.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/mincost/CapacityScalingMinimumCostFlow.java:
```diff
@@ -211,40 +211,27 @@ public V getFlowDirection(E edge) {
-            }
-            dualSolution = new DualSolution<>(dualVariables);
+        if(minimumCostFLow == null)
+            throw new RuntimeException("Cannot return a dual solution before getMinimumCostFlow(MinimumCostFlowProblem minimumCostFlowProblem) is invoked!");
```
This is too generic to throw. Either throw some standard exception that makes sense here, or maybe create some custom exception (which extends `RuntimeException`). 

As a more general solution (which arguably would be used also in other places) we could have a generic `JGraphTException` which extends RuntimeException and all others custom exceptions extend this one. For example we have something like `NegativeCycleDetectedException`. @jsichi Opinions?

jsichi replied:
IllegalStateException seems appropriate here.

I don't see much value in a JGraphTException for the kind of unchecked exceptions we currently throw.  It might make sense in the future if we ever introduce checked exceptions.

jkinable replied:
Actually, instead of throwing an exception, we might as well return null?
@return solution to the dual linear program formulated on the network, or null if no Minimum Cost Flow Problem has been solved.

Would you agree?

d-michail replied:
I try to avoid returning `null` these days. A similar solution is to return `Optional` but we are not using such a pattern in the library. Not sure whether we want to. @jsichi @Toptachamann Preferences?

jsichi replied:
I think it's fine if we start introducing Optional.

jkinable replied:
I really dislike using Optional here. Normally the user would invoke:
`Map<.> dualSol=getDual();`
Only when the user forgets to invoke `calculateMinimumCostFlow(.)` this method goes wrong.
However, by returning an optional, the simple `Map<.> dualSol=getDual();` now becomes:
```
Optional<Map<.>> optionalDualSol=getDual();
if(optionalDualSol.isPresent())
  dualSol=optionalDualSol.get();
```
This is so incredibly verbose, especially since in 99.9% of the cases this is not needed at all.

So I would strongly prefer to simply return null, which seems appropriate here, and let the user do an old-fashioned null check if necessary, or throw an IllegalStateException. @d-michail @jsichi what's your preference? I'd like to wrap this one up such that we can proceed with the release.

d-michail replied:
No strong preference. The documentation should contain the necessary information.

jkinable replied:
Ok, let's keep it as-is for now. @d-michail if satisfied, could you go ahead and merge?

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/mincost/CapacityScalingMinimumCostFlow.java:
```diff
@@ -161,31 +157,30 @@
+     */
+    @Override
+    public Map<E, Double> getFlowMap() {
+        return this.minimumCostFLow.getFlowMap();
```
We also need a null pointer check here.

jkinable commented:
@jsichi can you let us know which solution you prefer such that we can get this one merged? Without this PR I can't get the release out since I want to avoid the usual deprecation process.

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/MinimumCostFlowAlgorithm.java:
```diff
@@ -30,91 +32,52 @@
      * @return minimum cost flow
      */
-    MinimumCostFlow<E> getMinimumCostFlow();
+    MinimumCostFlow<E> getMinimumCostFlow(MinimumCostFlowProblem minimumCostFlowProblem);
```
Why don't we use the generic version of the `MinimumCostFlowProblem<V, E>`?

jkinable replied:
Oops, somehow missed this. Thanks for catching this!

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/interfaces/MinimumCostFlowAlgorithm.java:
```diff
@@ -30,91 +32,52 @@
      */
-    default double getFlowCost() {
-        return getMinimumCostFlow().getCost();
+    default double getFlowCost(MinimumCostFlowProblem minimumCostFlowProblem) {
```
Same here

Toptachamann commented on jgrapht-core/src/main/java/org/jgrapht/alg/flow/mincost/CapacityScalingMinimumCostFlow.java:
```diff
@@ -198,11 +197,16 @@ public V getFlowDirection(E edge) {
-    public MinimumCostFlow<E> getMinimumCostFlow() {
-        if (minimumCostFLow == null) {
-            lazyCalculateMinimumCostFlow();
+    public MinimumCostFlow<E> getMinimumCostFlow(final MinimumCostFlowProblem minimumCostFlowProblem) {
```
Same with the generic version

jsichi commented:
@jkinable is there still an unresolved issue that needs my input?  I saw a notification email to that effect, but I can't find it anywhere in the github UI.

d-michail commented:
I guess, it was triggered by Timofey's comments about the missing generic parameters.

jkinable commented:
Alright, fixed the last issue. If everybody is happy, please go ahead and merge this one :)

d-michail commented:
Great, thanks!
